dashboard/app/models/pd/post_course_survey.rb

```diff
@@ -13,12 +13,82 @@
 # Indexes
 #
 #  index_pd_post_course_surveys_on_form_id                (form_id)
+#  index_pd_post_course_surveys_on_form_user_year_course  (form_id,user_id,year,course) UNIQUE
 #  index_pd_post_course_surveys_on_submission_id          (submission_id) UNIQUE
 #  index_pd_post_course_surveys_on_user_form_year_course  (user_id,form_id,year,course) UNIQUE
 #  index_pd_post_course_surveys_on_user_id                (user_id)
 #
 
 module Pd
   class PostCourseSurvey < ActiveRecord::Base
+    include JotFormBackedForm
+
+    VALID_YEARS = [
+      YEAR_18_19 = '2018-2019'.freeze
+    ].freeze
+
+    VALID_COURSES = [
+      COURSE_CSD = 'CS Discoveries'.freeze,
+      COURSE_CSP = 'CS Principles'.freeze
+    ].freeze
+
+    CURRENT_YEAR = YEAR_18_19
+
+    belongs_to :user
+
+    # @override
+    def self.attribute_mapping
+      {
+        user_id: 'userId',
+        course: 'csp_or_csd'
+      }
+    end
+
+    validates_uniqueness_of :user_id, scope: [:form_id, :year, :course],
+      message: 'already has a submission for this form, year, and course'
+
+    validates_presence_of(
+      :user_id,
+      :course
+    )
+    validates_inclusion_of :year, in: VALID_YEARS
+    validates_inclusion_of :course, in: VALID_COURSES
+
+    # Skip other environments. Only keep this environment.
+    # @override
+    def skip_submission?(processed_answers)
+      environment = processed_answers['environment']
+      raise "Missing required environment field" unless environment
```

Comment:
Why raise the exception and not just skip (but log a warning)? Should this never happen?

Reply:
Correct, this should never happen. It would mean our embed is broken, or someone is getting around it and submitting JotForm responses without our required hidden fields.